### PR TITLE
Merge reset password features

### PR DESCRIPTION
**Pull request by ad08fee3**
Continued from https://github.com/gorel/dli-reports/pull/47 because merging wasn't pretty.
This branch adds password resetting and emailing. Admins cannot reset yet though.

### CODE REVIEW

File: dli_app/mod_auth/controllers.py [hunk 1]
```diff
@@ -103,3 +115,67 @@ def logout():
...
+        # Email is authenticated and sent
+        email = form.email.data
+	#Create User Specific URL here
+        pw_reset = PasswordReset(
+            user= User.get_by_email(email),
+          #  key=''.join(
+           #     random.choice(
+            #        string.ascii_letters + string.digits
+            #    ) for _ in range(60)
+           # ),
```
Comment (gorel) [inline dli_app/mod_auth/controllers.py]: Remove these comments.

File: dli_app/mod_auth/controllers.py [hunk 2]
```diff
@@ -3,6 +3,8 @@
 Author: Logan Gore
 This file is responsible for loading all site pages under /auth.
 """
+import random
+import string
```
Comment (gorel) [inline dli_app/mod_auth/controllers.py]: These imports are no longer used in this file

File: dli_app/mod_auth/controllers.py [hunk 3]
```diff
@@ -103,3 +115,67 @@ def logout():
...
+            #    ) for _ in range(60)
+           # ),
+        )
+        db.session.add(pw_reset)
+        db.session.commit()
+        #Send email here
+	mail = Mail(current_app)
+	title = 'Reset your Password'
+        url = "http://68.234.146.84:PORT /auth/setnewpass/"+pw_reset.key
+        content = 'Click this link to reset your password: '+url
```
Comment (gorel) [inline dli_app/mod_auth/controllers.py]: Put spaces around + operator

File: dli_app/mod_auth/controllers.py [hunk 4]
```diff
@@ -103,3 +115,67 @@ def logout():
...
+	sender = 'cs490testing@gmail.com'
+	msg = Message(title,sender=sender,recipients=[email])
```
Comment (gorel) [inline dli_app/mod_auth/controllers.py]: space after each comma
eg:
`Message(title, sender=sender, recipients=[email])`

File: dli_app/mod_auth/controllers.py [hunk 5]
```diff
@@ -103,3 +115,67 @@ def logout():
...
+	msg.body = content
+	mail.send(msg)
```
Comment (gorel) [inline dli_app/mod_auth/controllers.py]: You have a BUNCH of tab literal characters in here.
`:%s/\t/    /g`

File: dli_app/mod_auth/controllers.py [hunk 6]
```diff
@@ -103,3 +115,67 @@ def logout():
...
+        flash("Email sent!", "alert-success")
+        return redirect(url_for('default.home'))
+    else:
+        flash_form_errors(form)
+    	return render_template('auth/resetpass.html', form=form)
```
Comment (gorel) [inline dli_app/mod_auth/controllers.py]: Mixture of tab literals and spaces here

File: dli_app/mod_auth/controllers.py [hunk 7]
```diff
@@ -103,3 +115,67 @@ def logout():
...
+            return redirect(url_for('default.home'))
+        user.set_password(password)
+        db.session.commit()
+        flash("Password reset!", "alert-success")
+        return redirect(url_for('default.home'))
+    else:
+        flash_form_errors(form)
+        return render_template('auth/setnewpass.html', form=form, reset_key=reset_key)
+
+
```
Comment (gorel) [inline dli_app/mod_auth/controllers.py]: If this is the end of the file, we don't need these extra newlines

File: dli_app/mod_auth/forms.py [hunk 8]
```diff
@@ -212,3 +230,41 @@ class ForgotForm(Form):
             ),
         ],
     )
+
+
+class NewPassForm(Form):
+    """Form for resetting a user's password"""
+    def __init__(self, *args, **kwargs):
+        Form.__init__(self, *args, **kwargs)
+        self.user = None
```
Comment (gorel) [inline dli_app/mod_auth/forms.py]: self.user is never used for anything. Why is it here?

File: dli_app/mod_auth/forms.py [hunk 9]
```diff
@@ -212,3 +230,41 @@ class ForgotForm(Form):
...
+
+    def validate(self):
+        """Validate the form"""
+        if not Form.validate(self):
+            return False
+        return True
```
Comment (gorel) [inline dli_app/mod_auth/forms.py]: If all you do is call Form.validate, you don't need a validate() method

File: dli_app/mod_auth/models.py [hunk 10]
```diff
@@ -148,6 +152,31 @@ def get_by_email(cls, email):
         return User.query.filter_by(email=email).first()
 
 
+class PasswordReset(db.Model):
+    """Model for password reset key"""
+    id = db.Column(db.Integer, primary_key=True)
+    user_id= db.Column(db.Integer, db.ForeignKey('user.id'))
```
Comment (gorel) [inline dli_app/mod_auth/models.py]: Spaces around = operator

File: dli_app/mod_auth/models.py [hunk 11]
```diff
@@ -148,6 +152,31 @@ def get_by_email(cls, email):
...
+    key = db.Column(db.String(64))
+
+    def __init__(self, user, key=None):
+        """Initialize a  model"""
+        if key is None:
+            key =''.join(random.choice(string.ascii_letters + string.digits) for _ in range(60))
```
Comment (gorel) [inline dli_app/mod_auth/models.py]: Spaces around = operator

File: dli_app/mod_auth/models.py [hunk 12]
```diff
@@ -148,6 +152,31 @@ def get_by_email(cls, email):
...
+        self.user = user
+	self.key = key
+    def __repr__(self):
```
Comment (gorel) [inline dli_app/mod_auth/models.py:170]: Newline before function

File: dli_app/mod_auth/models.py [hunk 13]
```diff
@@ -148,6 +152,31 @@ def get_by_email(cls, email):
...
+    id = db.Column(db.Integer, primary_key=True)
+    user_id= db.Column(db.Integer, db.ForeignKey('user.id'))
+    key = db.Column(db.String(64))
+
+    def __init__(self, user, key=None):
+        """Initialize a  model"""
+        if key is None:
+            key =''.join(random.choice(string.ascii_letters + string.digits) for _ in range(60))
+        self.user = user
+	self.key = key
```
Comment (gorel) [inline dli_app/mod_auth/models.py]: Mix of tabs and spaces here, too

File: dli_app/mod_auth/models.py [hunk 14]
```diff
@@ -148,6 +152,31 @@ def get_by_email(cls, email):
...
+    def __repr__(self):
+	"""Return a descriptive representation of password reset"""
+	return '<Reset password for user %r>' % self.user
+    @classmethod
```
Comment (gorel) [inline dli_app/mod_auth/models.py:174]: Newline before function

File: dli_app/mod_auth/controllers.py (hunk 1 above)
Comment (ad08fee3) [inline dli_app/mod_auth/controllers.py]: Done.

File: dli_app/mod_auth/controllers.py (hunk 2 above)
Comment (ad08fee3) [inline dli_app/mod_auth/controllers.py]: Done.

File: dli_app/mod_auth/controllers.py (hunk 3 above)
Comment (ad08fee3) [inline dli_app/mod_auth/controllers.py]: Done.

File: dli_app/mod_auth/controllers.py (hunk 4 above)
Comment (ad08fee3) [inline dli_app/mod_auth/controllers.py]: Done.

File: dli_app/mod_auth/controllers.py (hunk 5 above)
Comment (ad08fee3) [inline dli_app/mod_auth/controllers.py]: I replaced all of the preceding tabs with spaces. Hopefully that's all of it.

File: dli_app/mod_auth/controllers.py (hunk 6 above)
Comment (ad08fee3) [inline dli_app/mod_auth/controllers.py]: I replaced all of the preceding tabs with spaces. Hopefully that's all of it.

File: dli_app/mod_auth/controllers.py (hunk 7 above)
Comment (ad08fee3) [inline dli_app/mod_auth/controllers.py]: Done.

File: dli_app/mod_auth/forms.py (hunk 8 above)
Comment (ad08fee3) [inline dli_app/mod_auth/forms.py]: Deleted.

File: dli_app/mod_auth/forms.py (hunk 9 above)
Comment (ad08fee3) [inline dli_app/mod_auth/forms.py]: Deleted.

File: dli_app/mod_auth/models.py (hunk 10 above)
Comment (ad08fee3) [inline dli_app/mod_auth/models.py]: Done.

File: dli_app/mod_auth/models.py (hunk 11 above)
Comment (ad08fee3) [inline dli_app/mod_auth/models.py]: Done.

File: dli_app/mod_auth/models.py (hunk 12 above)
Comment (ad08fee3) [inline dli_app/mod_auth/models.py:170]: Done.

File: dli_app/mod_auth/models.py (hunk 13 above)
Comment (ad08fee3) [inline dli_app/mod_auth/models.py]: I replaced all of the preceding tabs with spaces. Hopefully that's all of it.

File: dli_app/mod_auth/models.py (hunk 14 above)
Comment (ad08fee3) [inline dli_app/mod_auth/models.py:174]: Done.

**Comment by ad08fee3** [comment]
Added those quick fixes, and it still works. Merge it if it LGTM

**Pull request merged**